### PR TITLE
Dynamically load drivers

### PR DESCRIPTION
**Pull request by svmnotn**
Tasks:
- [ ] Names
- [ ] Structure
- [x] Segmentation Fault (code 11) **free**
- [x] Return Result instead of panicking
- [ ] Documented/'Good' API
- [x] Get tests to compile automatically in travis
- [ ] rebase with better commit names

### CODE REVIEW

File: src/loader.rs [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+use self::libloading::{Library, Symbol};
+use super::driver::{Driver, DriverSchema};
+use std::collections::HashMap;
+
+pub type Result<T> = ::std::io::Result<T>;
+
+/// Trait implemented by anthing that wishes to load a driver at runtime.
+pub trait DriverLoader {
+  fn get_driver<'a>(&self) -> Option<&'a Driver>;
+  fn get_schema<'a>(&self) -> Option<&'a DriverSchema>;
```
Comment (calebmer) [inline src/loader.rs]: Theoretically there will be more than these two traits. Like `DriverAccelerate` if we want to take that direction.

File: src/loader.rs [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+  pub fn new() -> Self {
+    Loader {
+      names: HashMap::new(),
+      libs: HashMap::new(),
+      drivers: HashMap::new()
+    }
+  }
+
+  fn get_loc(&self, name: String) -> String {
+    self.names.get(&name).unwrap().clone()
```
Comment (calebmer) [inline src/loader.rs]: Should we not use `unwrap` here?

File: src/loader.rs (hunk 1 above)
Comment (svmnotn) [inline src/loader.rs]: I know, I don't like it but from what I see we would have to add every possible trait there.
Also this is being re-written if you have not seen that none of the boxes are checked

File: src/loader.rs (hunk 2 above)
Comment (svmnotn) [inline src/loader.rs]: I can guarantee that it is only called with something that was loaded (Maybe, being re-written anyways)

File: src/loader.rs [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+  }
+
+  fn load_lib(&mut self, loc: String) {
+    let lib = match Library::new(loc.clone()) {
+      Ok(lib) => lib,
+      Err(e)  => panic!("Failed to load lib '{}' due to:\n{}", loc, e)
```
Comment (calebmer) [inline src/loader.rs]: What do you think about returning `Result` and handling errors in a root function vs. random `panic!`s? (`try!` may help here)

I'm not sure which is the better pattern.

Comment (svmnotn) [inline src/loader.rs]: read the tasks

File: src/loader.rs [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+  }
+
+  fn load_driver(&mut self, loc: String) {
+    if !self.libs.contains_key(&loc) {
+      self.load_lib(loc.clone());
+    }
+
+    let lib = self.libs.get(&loc).unwrap();
+
+    let get_loader: Symbol<extern fn() -> Box<DriverLoader>> = match unsafe { lib.get(b"get_loader\0") } {
```
Comment (calebmer) [inline src/loader.rs]: Oh no. The dark arts! 😊

File: src/loader.rs (hunk 3 above)
Comment (calebmer) [inline src/loader.rs]: Good idea.

File: src/loader.rs (hunk 1 above)
Comment (calebmer) [inline src/loader.rs]: I'll think if there's a better way to model multiple features. Any suggestions? Hopefully we can find a good solution for this part of the code and the driver bits.

File: src/structure.rs [hunk 5]
```diff
@@ -16,5 +16,5 @@ pub struct Structure<C: Collection> {
 pub trait Collection {
   /// Get the public name of the collection. In a PostgreSQL database, this
   /// would be the table name without the schema name.
-  fn get_name() -> String;
+  fn get_name(&self) -> String;
```
Comment (calebmer) [inline src/structure.rs]: Yes. My bad.

File: test-libs/good/.gitignore [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+Desktop.ini
+
+# Mac Files
+.DS_Store
+
+# Compiled files
+*.o
+*.so
+*.rlib
+*.dll
```
Comment (calebmer) [inline test-libs/good/.gitignore]: Do we really need all the `.gitignore`s?

Comment (svmnotn) [inline test-libs/good/.gitignore]: try it without them.

Comment (calebmer) [inline test-libs/good/.gitignore]: `target` isn't ignored right? That's because you have `/target/` in the `.gitignore`.

Comment (svmnotn) [inline test-libs/good/.gitignore]: I'll test but it was adding all those files

Comment (calebmer) [inline test-libs/good/.gitignore]: Yeah. Try just `target` without the slashes.

Comment (svmnotn) [inline test-libs/good/.gitignore]: nope, once I remove the .gitignore it adds all the files

Comment (calebmer) [inline test-libs/good/.gitignore]: That's ok. I fixed it.

File: test-libs/good/src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+use ardite_core::driver::{Driver, DriverSchema};
+
+#[no_mangle]
+pub extern fn get_loader() -> Box<DriverLoader> {
+  println!("GETTING LOADER!");
+  Box::new(Loader {s: "test".to_string()})
+}
+
+#[no_mangle]
+pub extern fn eat_loader(driver: Box<DriverLoader>) {
```
Comment (calebmer) [inline test-libs/good/src/lib.rs]: Eat loader 😊

Comment (svmnotn) [inline test-libs/good/src/lib.rs]: sorry tis gone. I think I fixed the seg fault tho :grinning:

**Comment by svmnotn** [comment]
... this is weird. It builds just fine on my machine

**Comment by calebmer** [comment]
Mmm. The dark arts of unsafe rust?

**Comment by svmnotn** [comment]
can you try building it on your computer?

**Comment by calebmer** [comment]
Do I just run `scripts/build-tests.sh`?

**Comment by calebmer** [comment]
Also, how familiar are you with `Makefile`s?

**Comment by svmnotn** [comment]
you have to modify it since you are in mac, and I've seen them, not use tho

**Comment by calebmer** [comment]
I tried `TRAVIS_OS_NAME=osx scripts/build-tests.sh`. Nothing crashed.

**Comment by svmnotn** [comment]
ok good I guess

**Comment by calebmer** [comment]
Actually wait. Running `cargo test --verbose` gives me this:

```
       Fresh lazy_static v0.1.15
       Fresh winapi v0.2.5
       Fresh winapi-build v0.1.1
       Fresh kernel32-sys v0.2.1
       Fresh libloading v0.2.1
       Fresh ardite-core v0.0.1 (file:///Users/calebmer/Projects/ardite/ardite-core)
     Running `/Users/calebmer/Projects/ardite/ardite-core/target/debug/ardite_core-ce4abb233c20466d`

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured

     Running `/Users/calebmer/Projects/ardite/ardite-core/target/debug/lib-7030db084623d5d7`

running 3 tests
test load_missing ... ok
test load_missing_fn ... ok
GETTING LOADER!
Process didn't exit successfully: `/Users/calebmer/Projects/ardite/ardite-core/target/debug/lib-7030db084623d5d7` (signal: 4)

Caused by:
  Process didn't exit successfully: `/Users/calebmer/Projects/ardite/ardite-core/target/debug/lib-7030db084623d5d7` (signal: 4)
```

**Comment by svmnotn** [comment]
That is funny... because I get

```
[svmnotn@uss-shadow ardite-core]$ TRAVIS_OS_NAME=linux scripts/build-tests.sh 
building libgood
   Compiling winapi v0.2.5
   Compiling libloading v0.2.1
   Compiling lazy_static v0.1.15
   Compiling winapi-build v0.1.1
   Compiling kernel32-sys v0.2.1
   Compiling ardite-core v0.0.1 (file:///home/svmnotn/data/dev/ardite-core/test-libs/good)
   Compiling good v0.1.0 (file:///home/svmnotn/data/dev/ardite-core/test-libs/good)
moved libgood
building libmissing
   Compiling missing v0.1.0 (file:///home/svmnotn/data/dev/ardite-core/test-libs/missing)
moved libmissing
[svmnotn@uss-shadow ardite-core]$ cargo build
   Compiling winapi-build v0.1.1
   Compiling winapi v0.2.5
   Compiling libloading v0.2.1
   Compiling lazy_static v0.1.15
   Compiling kernel32-sys v0.2.1
   Compiling ardite-core v0.0.1 (file:///home/svmnotn/data/dev/ardite-core)
[svmnotn@uss-shadow ardite-core]$ cargo test
   Compiling ardite-core v0.0.1 (file:///home/svmnotn/data/dev/ardite-core)
     Running target/debug/ardite_core-a2faac789214442f

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured

     Running target/debug/lib-c70371c2d5036cae

running 3 tests
test load_missing ... ok
GETTING LOADER!
test load_good ... ok
test load_missing_fn ... ok

test result: ok. 3 passed; 0 failed; 0 ignored; 0 measured

   Doc-tests ardite-core

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured
```

**Comment by svmnotn** [comment]
are you running the latest?

**Comment by calebmer** [comment]
Just made your script work without an env variable and reduced duplication. I'll run the tests again 😊

**Comment by svmnotn** [comment]
You didn't push the fixed script... also you realize that the lib endings are different right?
Or at least they should be

**Comment by calebmer** [comment]
Here is bash on the latest version:

```
Calebs-MacBook-Pro:ardite-core calebmer$ scripts/build-tests.sh && cargo test --verbose
Building libgood
   Compiling libloading v0.2.1
   Compiling winapi-build v0.1.1
   Compiling lazy_static v0.1.15
   Compiling winapi v0.2.5
   Compiling kernel32-sys v0.2.1
   Compiling ardite-core v0.0.1 (file:///Users/calebmer/Projects/ardite/ardite-core/test-libs/good)
   Compiling good v0.1.0 (file:///Users/calebmer/Projects/ardite/ardite-core/test-libs/good)
Building libmissing
   Compiling missing v0.1.0 (file:///Users/calebmer/Projects/ardite/ardite-core/test-libs/missing)
Moved libgood
Moved libmissing
       Fresh lazy_static v0.1.15
       Fresh winapi v0.2.5
       Fresh winapi-build v0.1.1
       Fresh kernel32-sys v0.2.1
       Fresh libloading v0.2.1
       Fresh ardite-core v0.0.1 (file:///Users/calebmer/Projects/ardite/ardite-core)
     Running `/Users/calebmer/Projects/ardite/ardite-core/target/debug/ardite_core-ce4abb233c20466d`

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured

     Running `/Users/calebmer/Projects/ardite/ardite-core/target/debug/lib-7030db084623d5d7`

running 3 tests
GETTING LOADER!
test load_missing ... ok
Process didn't exit successfully: `/Users/calebmer/Projects/ardite/ardite-core/target/debug/lib-7030db084623d5d7` (signal: 4)

Caused by:
  Process didn't exit successfully: `/Users/calebmer/Projects/ardite/ardite-core/target/debug/lib-7030db084623d5d7` (signal: 4)
```

**Comment by calebmer** [comment]
I did [push](https://github.com/ardite/ardite-core/blob/feature/driver/scripts/build-tests.sh) the script.

**Comment by svmnotn** [comment]
... I have no clue what is going on at this point... and I can't test it....

**Comment by svmnotn** [comment]
Tested on Windows 7 64bit, no segfault found.
To note:

This happened on a clean install of Rust + a just downloaded version of the latest from this branch with some changes to improve code reuse: 90362fc

**Comment by svmnotn** [comment]
Tests also pass on a clean install of rust + the latest commit, on a 32bit mint install

**Comment by calebmer** [comment]
On a clean copy as requested:

```
Calebs-MacBook-Pro:ardite-core-feature-driver calebmer$ scripts/build-tests.sh && cargo test --verbose
Building libgood
    Updating registry `https://github.com/rust-lang/crates.io-index`
   Compiling lazy_static v0.1.15
   Compiling winapi v0.2.5
   Compiling winapi-build v0.1.1
   Compiling libloading v0.2.1
   Compiling kernel32-sys v0.2.1
   Compiling ardite-core v0.0.1 (file:///Users/calebmer/Downloads/ardite-core-feature-driver/test-libs/good)
   Compiling good v0.1.0 (file:///Users/calebmer/Downloads/ardite-core-feature-driver/test-libs/good)
Building libmissing
   Compiling missing v0.1.0 (file:///Users/calebmer/Downloads/ardite-core-feature-driver/test-libs/missing)
Moved libgood
Moved libmissing
    Updating registry `https://github.com/rust-lang/crates.io-index`
   Compiling winapi-build v0.1.1
     Running `rustc /Users/calebmer/.cargo/registry/src/github.com-0a35038f75765ae4/winapi-build-0.1.1/src/lib.rs --crate-name build --crate-type lib -g -C metadata=304afb6bdff23d72 -C extra-filename=-304afb6bdff23d72 --out-dir /Users/calebmer/Downloads/ardite-core-feature-driver/target/debug/deps --emit=dep-info,link -L dependency=/Users/calebmer/Downloads/ardite-core-feature-driver/target/debug/deps -L dependency=/Users/calebmer/Downloads/ardite-core-feature-driver/target/debug/deps --cap-lints allow`
   Compiling winapi v0.2.5
     Running `rustc /Users/calebmer/.cargo/registry/src/github.com-0a35038f75765ae4/winapi-0.2.5/src/lib.rs --crate-name winapi --crate-type lib -g -C metadata=d88a8c018c340227 -C extra-filename=-d88a8c018c340227 --out-dir /Users/calebmer/Downloads/ardite-core-feature-driver/target/debug/deps --emit=dep-info,link -L dependency=/Users/calebmer/Downloads/ardite-core-feature-driver/target/debug/deps -L dependency=/Users/calebmer/Downloads/ardite-core-feature-driver/target/debug/deps --cap-lints allow`
   Compiling lazy_static v0.1.15
     Running `rustc /Users/calebmer/.cargo/registry/src/github.com-0a35038f75765ae4/lazy_static-0.1.15/src/lib.rs --crate-name lazy_static --crate-type lib -g -C metadata=f3aa6dfcc7c157cc -C extra-filename=-f3aa6dfcc7c157cc --out-dir /Users/calebmer/Downloads/ardite-core-feature-driver/target/debug/deps --emit=dep-info,link -L dependency=/Users/calebmer/Downloads/ardite-core-feature-driver/target/debug/deps -L dependency=/Users/calebmer/Downloads/ardite-core-feature-driver/target/debug/deps --cap-lints allow`
   Compiling libloading v0.2.1
     Running `rustc /Users/calebmer/.cargo/registry/src/github.com-0a35038f75765ae4/libloading-0.2.1/build.rs --crate-name build_script_build --crate-type bin -g --out-dir /Users/calebmer/Downloads/ardite-core-feature-driver/target/debug/build/libloading-6f9f475745955a63 --emit=dep-info,link -L dependency=/Users/calebmer/Downloads/ardite-core-feature-driver/target/debug/deps -L dependency=/Users/calebmer/Downloads/ardite-core-feature-driver/target/debug/deps --cap-lints allow`
     Running `/Users/calebmer/Downloads/ardite-core-feature-driver/target/debug/build/libloading-6f9f475745955a63/build-script-build`
   Compiling kernel32-sys v0.2.1
     Running `rustc /Users/calebmer/.cargo/registry/src/github.com-0a35038f75765ae4/kernel32-sys-0.2.1/build.rs --crate-name build_script_build --crate-type bin -g --out-dir /Users/calebmer/Downloads/ardite-core-feature-driver/target/debug/build/kernel32-sys-35c1a745c861d23a --emit=dep-info,link -L dependency=/Users/calebmer/Downloads/ardite-core-feature-driver/target/debug/deps -L dependency=/Users/calebmer/Downloads/ardite-core-feature-driver/target/debug/deps --extern build=/Users/calebmer/Downloads/ardite-core-feature-driver/target/debug/deps/libbuild-304afb6bdff23d72.rlib --cap-lints allow`
     Running `/Users/calebmer/Downloads/ardite-core-feature-driver/target/debug/build/kernel32-sys-35c1a745c861d23a/build-script-build`
     Running `rustc /Users/calebmer/.cargo/registry/src/github.com-0a35038f75765ae4/kernel32-sys-0.2.1/src/lib.rs --crate-name kernel32 --crate-type lib -g -C metadata=35c1a745c861d23a -C extra-filename=-35c1a745c861d23a --out-dir /Users/calebmer/Downloads/ardite-core-feature-driver/target/debug/deps --emit=dep-info,link -L dependency=/Users/calebmer/Downloads/ardite-core-feature-driver/target/debug/deps -L dependency=/Users/calebmer/Downloads/ardite-core-feature-driver/target/debug/deps --extern winapi=/Users/calebmer/Downloads/ardite-core-feature-driver/target/debug/deps/libwinapi-d88a8c018c340227.rlib --cap-lints allow`
     Running `rustc /Users/calebmer/.cargo/registry/src/github.com-0a35038f75765ae4/libloading-0.2.1/src/lib.rs --crate-name libloading --crate-type lib -g -C metadata=6f9f475745955a63 -C extra-filename=-6f9f475745955a63 --out-dir /Users/calebmer/Downloads/ardite-core-feature-driver/target/debug/deps --emit=dep-info,link -L dependency=/Users/calebmer/Downloads/ardite-core-feature-driver/target/debug/deps -L dependency=/Users/calebmer/Downloads/ardite-core-feature-driver/target/debug/deps --extern winapi=/Users/calebmer/Downloads/ardite-core-feature-driver/target/debug/deps/libwinapi-d88a8c018c340227.rlib --extern lazy_static=/Users/calebmer/Downloads/ardite-core-feature-driver/target/debug/deps/liblazy_static-f3aa6dfcc7c157cc.rlib --extern kernel32=/Users/calebmer/Downloads/ardite-core-feature-driver/target/debug/deps/libkernel32-35c1a745c861d23a.rlib --cap-lints allow`
   Compiling ardite-core v0.0.1 (file:///Users/calebmer/Downloads/ardite-core-feature-driver)
     Running `rustc src/lib.rs --crate-name ardite_core --crate-type lib -g --test -C metadata=d1a49f0d6a34a3c8 -C extra-filename=-d1a49f0d6a34a3c8 --out-dir /Users/calebmer/Downloads/ardite-core-feature-driver/target/debug --emit=dep-info,link -L dependency=/Users/calebmer/Downloads/ardite-core-feature-driver/target/debug -L dependency=/Users/calebmer/Downloads/ardite-core-feature-driver/target/debug/deps --extern libloading=/Users/calebmer/Downloads/ardite-core-feature-driver/target/debug/deps/liblibloading-6f9f475745955a63.rlib`
     Running `rustc src/lib.rs --crate-name ardite_core --crate-type lib -g --out-dir /Users/calebmer/Downloads/ardite-core-feature-driver/target/debug --emit=dep-info,link -L dependency=/Users/calebmer/Downloads/ardite-core-feature-driver/target/debug -L dependency=/Users/calebmer/Downloads/ardite-core-feature-driver/target/debug/deps --extern libloading=/Users/calebmer/Downloads/ardite-core-feature-driver/target/debug/deps/liblibloading-6f9f475745955a63.rlib`
     Running `rustc tests/lib.rs --crate-name lib --crate-type bin -g --test -C metadata=e152671c4628625d -C extra-filename=-e152671c4628625d --out-dir /Users/calebmer/Downloads/ardite-core-feature-driver/target/debug --emit=dep-info,link -L dependency=/Users/calebmer/Downloads/ardite-core-feature-driver/target/debug -L dependency=/Users/calebmer/Downloads/ardite-core-feature-driver/target/debug/deps --extern libloading=/Users/calebmer/Downloads/ardite-core-feature-driver/target/debug/deps/liblibloading-6f9f475745955a63.rlib --extern ardite_core=/Users/calebmer/Downloads/ardite-core-feature-driver/target/debug/libardite_core.rlib`
     Running `/Users/calebmer/Downloads/ardite-core-feature-driver/target/debug/ardite_core-d1a49f0d6a34a3c8`

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured

     Running `/Users/calebmer/Downloads/ardite-core-feature-driver/target/debug/lib-e152671c4628625d`

running 3 tests
GETTING LOADER!
Process didn't exit successfully: `/Users/calebmer/Downloads/ardite-core-feature-driver/target/debug/lib-e152671c4628625d` (signal: 4)

Caused by:
  Process didn't exit successfully: `/Users/calebmer/Downloads/ardite-core-feature-driver/target/debug/lib-e152671c4628625d` (signal: 4)
```

**Comment by svmnotn** [comment]
Making a new PR based on the latest changes, as they add some requirements that this would not support (also the git history could be better)

**Comment by calebmer** [comment]
:cry:

**Comment by calebmer** [comment]
Wait, one commit wasn't good enough? 😊

**Pull request closed without merging**